apps/judicial-system/web/src/graphql/gql.ts

```diff
@@ -0,0 +1,353 @@
+      custodyEndDate
+      custodyRestrictions
+      accusedAppealDecision
+      accusedAppealAnnouncement
+      prosecutorAppealDecision
+      prosecutorAppealAnnouncement
+      # judgeId
+      judge {
+        nationalId
+        name
+        title
+        mobileNumber
+        email
+        role
+      }
+      # notifications
+    }
+  }
+`
+
+export const CaseQuery = gql`
+  query CaseQuery($id: ID!) {
+    case(id: $id) {
+      id
+      created
+      modified
+      state
+      policeCaseNumber
+      accusedNationalId
+      accusedName
+      accusedAddress
+      court
+      arrestDate
+      requestedCourtDate
+      requestedCustodyEndDate
+      lawsBroken
+      custodyProvisions
+      requestedCustodyRestrictions
+      caseFacts
+      witnessAccounts
+      investigationProgress
+      legalArguments
+      comments
+      # prosecutorId
+      prosecutor {
+        nationalId
+        name
+        title
+        mobileNumber
+        email
+        role
+      }
+      courtCaseNumber
+      courtStartTime
+      courtEndTime
+      courtAttendees
+      policeDemands
+      accusedPlea
+      litigationPresentations
+      ruling
+      rejecting
+      custodyEndDate
+      custodyRestrictions
+      accusedAppealDecision
+      accusedAppealAnnouncement
+      prosecutorAppealDecision
+      prosecutorAppealAnnouncement
+      # judgeId
+      judge {
+        nationalId
+        name
+        title
+        mobileNumber
+        email
+        role
+      }
+      # notifications
+    }
+  }
+`
+
+export const CreateCaseMutation = gql`
+  mutation CreateCaseMutation($createCase: CreateCaseDto!) {
+    createCase(createCase: $createCase) {
+      id
+      created
+      modified
+      state
+      policeCaseNumber
+      accusedNationalId
+      accusedName
+      accusedAddress
+      court
+      arrestDate
+      requestedCourtDate
+      requestedCustodyEndDate
+      lawsBroken
+      custodyProvisions
+      requestedCustodyRestrictions
+      caseFacts
+      witnessAccounts
+      investigationProgress
+      legalArguments
+      comments
+      # prosecutorId
+      prosecutor {
+        nationalId
+        name
+        title
+        mobileNumber
+        email
+        role
+      }
+      courtCaseNumber
+      courtStartTime
+      courtEndTime
+      courtAttendees
+      policeDemands
+      accusedPlea
+      litigationPresentations
+      ruling
+      rejecting
+      custodyEndDate
+      custodyRestrictions
+      accusedAppealDecision
+      accusedAppealAnnouncement
+      prosecutorAppealDecision
+      prosecutorAppealAnnouncement
+      # judgeId
+      judge {
+        nationalId
+        name
+        title
+        mobileNumber
+        email
+        role
+      }
+      # notifications
+    }
+  }
+`
+
+export const UpdateCaseMutation = gql`
+  mutation UpdateCaseMutation($id: ID!, $updateCase: UpdateCaseDto!) {
+    updateCase(id: $id, updateCase: $updateCase) {
+      id
```

Comment:
Why are the queries/mutations split up from the components?

This most often leads to objects being requested that aren't needed in the components (looks like these mutations are requesting all the objects).

Each component should only request what they need.

Reply:
Not sure I understand this comment. We are only returning the updated object.

Reply:
e.g. in `apps/judicial-system/web/src/routes/DetentionRequests/DetentionRequests.tsx` you're only working with 
```
      id
      policeCaseNumber
      accusedName
      accusedNationalId
      state
      custodyEndDate
```
.. but that query is requesting:
```typescript
      id
      created
      modified
      state
      policeCaseNumber
      accusedNationalId
      accusedName
      accusedAddress
      court
      arrestDate
      requestedCourtDate
      requestedCustodyEndDate
      lawsBroken
      custodyProvisions
      requestedCustodyRestrictions
      caseFacts
      witnessAccounts
      investigationProgress
      legalArguments
      comments
      # prosecutorId
      prosecutor {
        nationalId
        name
        title
        mobileNumber
        email
        role
      }
      courtCaseNumber
      courtStartTime
      courtEndTime
      courtAttendees
      policeDemands
      accusedPlea
      litigationPresentations
      ruling
      rejecting
      custodyEndDate
      custodyRestrictions
      accusedAppealDecision
      accusedAppealAnnouncement
      prosecutorAppealDecision
      prosecutorAppealAnnouncement
      # judgeId
      judge {
        nationalId
        name
        title
        mobileNumber
        email
        role
      }
      # notifications
    }
```
which is a bloated payload and not what the component needs.
Here, the `GraphQL` is being used like `REST`, removing the power of GraphQL.

Same for the `mutations`. What is it that the individual component is about to change using the mutation? The component should then only request the fields that are about to be changed.

If you specify the `GraphQL ID` (like mentioned in previous comments) then the Apollo store will respectfully update the corresponding objects that are later requested with additional data.

Reply:
Your comment on the query used in DetentionRequests makes sense. I will limit the fields returned to what is actually used.

Regarding the mutations and GraphQL automatically (respectfully :) ) updating its store, this makes sense as long as we have a single user working on the same court case. That is the situation we expect for now, but later, we will likely have multiple people working on the same court case and need to guard against that. In such cases, we will need to be able to ask Apollo to invalidate its cache or not use the cache at all. Do you have any thoughts/pointers regarding this?

Reply:
Yeah, best case scenario would be to setup GraphQL Subscription instead of Queries -> over websocket

Reply:
Right, we fight that battle in due time.

Reply:
I have now addressed most of the issues you mentioned here.

One thing I was struggling with and did not address was updates to court cases. In most cases the client updates a single field at a time or when the corresponding input field looses focus. Ideally, I would wan the mutation to query for the exact same fields that are being updated, plus some select other fields, such as a modified timestamp.

This would something like this:

mutation UpdateCaseMutation($input: UpdateCaseInput!) {
    updateCase(input: $input) {
      id
      modified
      ...$input
    }
}

However, I don't know how to create such a query, except perhaps by using dynamic queries generated at run time.

So, I will submit this PR for another review. I would appreciate any suggestions as to how I can handle these update mutations. Alternatively, it would be great if you would approve the PR and we would improve this later.

Reply:
I believe the main issue here is that `updateCase` is too broad. It doesn't specify "what" it's updating, going against the GraphQL Mutation principal: `Make mutations as specific as possible.`.
There you end up with the same problem as before, querying all.
Moving the AST to the component would also document quite well what the component is doing, e.g.
```graphql
query JudicialSystemUserQuery {
  judicialSystemUser {
    id
    name
    address
  }
}

mutation JudicialSystemChangeAddressMutation($input: JudicialSystemChangeAddressInput) {
  judicialSystemChangeAddress(input: $input) {
    id
    address
  }
}
```

You can even batch up the mutations, to save round-trips (mutations run in series):
```graphql
mutation JudicialSystemUpdateUserMutation($input: JudicialSystemUpdateUserInput) {
  judicialSystemChangeAddress(..) {
    id
    address
  }
  judicialSystemChangeName(..) {
    id
    name
  }
}
```

Reply:
I understand the issue, but with this approach I will end up with a lot of mutations. One for each single field update and one for any combination of multi field updates possible.

The general rule in the UI is to save each field separately, but there are a few cases where more than one field are saved at the same time.

Later on, we will also need to deal with the case where some of the mutations will fail, because of network issues and such. Then, we will try to save all edited, but not yet saved fields. Creating a hard coded mutation for each case is not feasible. 

So my question is rather: How do people handle such cases in general? Do they generate dynamic mutations at run time?

Reply:
I'll do some experiments, but what is your opinion on dynamically creating mutations at run-time?

Reply:
Yeah I see :/.. I've had a similar problem which I didn't look any deeper into.
The dynamic field fetching sounds like a good solution!
Having a top-level function that returns GraphQL AST :+1:

Reply:
👍

Reply:
Turns out we don't want Apollo to cache the queries. This is because more than one user can be modifying/creating court cases at the same time and we want the latest data/updates to be loaded when queries are executed. This means that we really don't need the update mutation to return any data, so I modified the mutation to only return a "modified" timestamp.

Reply:
Would be great to get another review.